### ISSUE TITLE
[Bug][ROCm] QuickReduce converts bf16 to fp16 without saturation: activations above 65504 silently become inf

### ISSUE THREAD

**Issue by stefanskiasan** [vllm-project/vllm#55345]
### Your current environment

ROCm 7.2.3, MI355X (gfx950), TP2 and TP4, vLLM main.
QuickReduce enabled with `VLLM_ROCM_QUICK_REDUCE_QUANTIZATION=INT8`, bf16 model.

### 🐛 Describe the bug

QuickReduce converts bf16 inputs to fp16 before packing them, and the
conversion does not saturate. bf16 covers the fp32 exponent range (up to
~3.4e38); fp16 stops at 65504. Any activation above that becomes `inf` and
poisons the reduction — silently, with no warning and no numerical check.

`csrc/quickreduce/quick_reduce_impl.cuh` (line numbers from current main):

```cpp
        const nv_bfloat162* bf_buf = reinterpret_cast<const nv_bfloat162*>(&tA[i]);
        half2 half_buf[4];
#pragma unroll
        for (int j = 0; j < 4; ++j) {
          float2 f = __bfloat1622float2(bf_buf[j]);   // 740
          half_buf[j] = __float22half2_rn(f);         // 741  > 65504 -> inf
        }
```

This sits on the path taken for **every** codec (INT4/INT6/INT8 and FP), not
just the quantized ones — the fp16 packing happens before the codec runs.

### Reproduction

Two ranks, INT8 codec, a single out-of-fp16-range value on rank 0 only, all
other elements 1.0:

```python
x = torch.full((n,), 1.0, dtype=torch.bfloat16, device=dev)
if rank == 0:
    x[0] = 1.0e30           # representable in bf16, not in fp16
out = qr.quick_all_reduce(x)
print(out[0].float().item(), torch.isfinite(out).all().item())
```

```
current main : inf     False
with clamping: 65536   True      # other elements unchanged at 2.0 in both runs
```

Same container image in both runs, only the built extension differs, so the
difference is attributable to the conversion and nothing else.

### Suggested fix

Clamp before converting — three lines, no measurable cost on the critical path
(the surrounding loop is bandwidth-bound):

```cpp
static constexpr float kFp16Max = 65504.0f;
...
          float2 f = __bfloat1622float2(bf_buf[j]);
          f.x = fminf(fmaxf(f.x, -kFp16Max), kFp16Max);
          f.y = fminf(fmaxf(f.y, -kFp16Max), kFp16Max);
          half_buf[j] = __float22half2_rn(f);
```

**This is a mitigation, not a complete fix, and I want to be precise about the
limit:** clamping stops a single large value from turning the whole line into
`inf`, but the *sum* can still overflow — with N ranks each contributing a
clamped 65504, the fp16 accumulation saturates again. A complete fix would keep
bf16 end to end for bf16 inputs rather than routing through fp16. Clamping is
the cheap part that removes the silent-`inf` failure mode; whether the
end-to-end bf16 path is worth its cost is a separate question.

Happy to send the clamping patch as a PR if that direction is acceptable.

**Comment by stefanskiasan** [comment]
Concrete end-to-end evidence for this, on GLM-5.3 (full model, `glm_moe_dsa`) on 8× MI355X (gfx950), ROCm 7.2.3, TP4, MTP k=4, bf16 activations.

We compared `VLLM_ROCM_QUICK_REDUCE_QUANTIZATION=INT8` against `INT4` — nothing else changed, both arms with prefix caching off.

**Throughput and short-prompt accuracy look like a free win for INT4:**

| | throughput | GSM8K (full 1319) | TPOT |
|---|---|---|---|
| INT8 | 1316.6 tok/s | 1273/1319 = 96.5 % | 12.39 ms |
| INT4 | **1365.2 tok/s (+3.7 %)** | **1279/1319 = 97.0 %** | 12.40 ms |

**At long context it is not.** Needle-in-a-haystack, five needles, 366 709-token prompt:

| run | INT8 | INT4 |
|---|---|---|
| 1 | **5/5** | 4/5 |
| 2 | — | **2/5** |
| 3 | — | **1/5** |

The failure mode is the interesting part. These are not retrieval misses — the model **finds** the needle and misreads it, one digit at a time:

```
needle 91274  ->  model answers "91244"
needle 40863  ->  model answers "40803"
```

That is what a value corruption looks like, not a lookup failure. At 92k tokens INT4 is still clean (5/5); the damage appears only once the all-reduce messages get large, which is exactly the regime this issue is about.

Two practical notes for anyone else measuring this:

1. **GSM8K does not catch it.** Its prompts are short, so the large messages never occur — INT4 even scored slightly *better* there. An acceptance test that misses the affected code path proves nothing.
2. **Check the effective value with `printenv` inside the container**, not by filtering `docker inspect`. Our env file ended up with both `...=INT8` (from the base) and `...=INT4` (appended); `grep` returns the first, the process sees the last. We briefly discarded a valid measurement over this.

We are staying on INT8 in production.

**Comment by stefanskiasan** [comment]
Follow-up with a cleaner measurement of the *user-visible* effect, and how this bug connects to two others.

Setting `VLLM_ROCM_QUICK_REDUCE_CAST_BF16_TO_FP16=0` cut our rate of corrupted responses under concurrent load from **2.0 % to 0.75 %**, measured on 2120 requests with an unchanged workload. The saturation fix in #55346 should achieve the same without giving up the fp16 speedup.

**How the inf becomes visible output.** The `inf` produced here does not stay an `inf`. It propagates into a logits row, and with speculative decoding the rejection sampler turns that row into a token: `VOCAB_BLOCK_SIZE = 8192`, NaN block maxima are mapped to `-inf` (#50183), so when the entire row is NaN, block 0 wins and its corrupted local argmax yields id **8191** — a perfectly valid token. The response then breaks off mid-word and degenerates into digit soup:

```
'... ist **Tokiohandle,0,0.1,112. the0322 3n.0\nThe2222...'
```

So the failure is silent: no assert, no error, no metric. Details in #53029.

**This was one of three independent inf/NaN sources on the same path.** Cumulative measurements, same detector and prompts throughout:

| configuration | requests | corrupted | rate |
|---|---|---|---|
| baseline | ~150 | 3 | 2.0 % |
| **+ cast disabled (this issue)** | 2120 | 16 | **0.75 %** |
| + QuickReduce acquire fence in `wait_sync_flag` | 880 | 3 | 0.34 % |
| + zero-init of the ROCm MLA output buffer (#37682) | 1040 | 0 | 0 % |

Two side notes that may be useful:

- Disabling the cast also *improved* speculative decoding: draft acceptance rose from 24–28 % to 41 %. Fewer poisoned values simply mean better draft predictions, so the fix is not a pure trade-off against the fp16 speedup.
- The second row above is a separate bug in the same file: `wait_sync_flag` in `csrc/quickreduce/base.h` spins on `__ATOMIC_RELAXED` with no acquire fence afterwards, so a rank can read a partner's payload before its flag. Adding an acquire fence on the wait side plus a release fence on the send side accounted for the 0.75 % → 0.34 % step. I can open a separate issue for that if it is useful.
